### PR TITLE
Add hostname validation to api-controller

### LINKED ISSUE #386: Verify uniqueness of hostnames in API resources in api-controller
Confirm these statements before you submit the issue:

- [x] I have searched open and closed issues for duplicates.
- [x] I have read the contributing guidelines.
---

**Description**

Implement validation of hostnames uniqueness in API resources in api-controller and set status invalid in case of duplicated hostname.

**Reasons**

In the API resources you are defining hostnames under which the API will be accessible. Then api-controller creates Virtual Service, which is an Istio resource used, among others, to expose an app on Gateway under some hostname. Currently, Istio delegates to a user the responsibility to ensure that the blocks of routes in two virtual services don't conflict with each other. In our case, it will always be conflicting as we are exposing services on host and match requests on every path by using "/.*" regex. It may lead to service exposure not working while a user doesn't get any feedback.

**Attachments**

### PR DESCRIPTION
**Pull request by kubadz**
Confirm these statements before you submit your pull request:

- [x] I have read and submitted the required [Contributor Licence Agreements](https://github.com/kyma-project/community/blob/master/CONTRIBUTING.md#agreements-and-licenses).
- [x] This pull request follows the contributing guidelines, recommended Git workflow, and templates.
- [x] I have tested my changes.
- [x] I have updated the relevant documentation.
---

**Description**

Changes proposed in this pull request:

- Implement validation of hostnames uniqueness in API resources in api-controller and set status invalid in case of duplicated hostname.
- Add info about required environment variable "DOMAIN_NAME" to the readme file

**Issue link**

Resolves #386

### CODE REVIEW

File: components/api-controller/README.md [hunk 1]
```diff
@@ -37,12 +37,13 @@ $ go run cmd/controller/main.go
...
-| API_CONTROLLER_LOG_LEVEL | No | `info` | Show detailed logs in the application.
-| DEFAULT_ISSUER | Yes | - | Used to set default issuer in NetworkPolicy
-| DEFAULT_JWKS_URI | Yes | - | Used to set default jwksUri in NetworkPolicy
-| GATEWAY_FQDN | Yes | - | Used to set gateway in VirtualServices spec
+| Name | Required | Default | Description | Possible Values |
+|------|----------|---------|-------------|-----------------|
+| API_CONTROLLER_LOG_LEVEL | No | `info` | Show detailed logs in the application. | `info`, `debug`
+| DEFAULT_ISSUER | Yes | - | Used to set default issuer in NetworkPolicy | any string |
+| DEFAULT_JWKS_URI | Yes | - | Used to set default jwksUri in NetworkPolicy | any string |
+| GATEWAY_FQDN | Yes | - | Used to set gateway in VirtualServices spec | any string |
```
Comment (klaudiagrz) [inline components/api-controller/README.md]: spec -> `specification` ?
`Used to set Gateway in the VirtualServices specification.`

File: components/api-controller/README.md [hunk 2]
```diff
@@ -37,12 +37,13 @@ $ go run cmd/controller/main.go
...
+| DOMAIN_NAME | Yes | - | Used to set hostname in VirtualServices spec when short version of hostname is provided | any string |
```
Comment (klaudiagrz) [inline components/api-controller/README.md]: spec -> specification ?
`Used to set a hostname in the VirtualServices specification if a short version of the hostname is provided.`

File: components/api-controller/pkg/apis/gateway.kyma.cx/meta/v1/types.go [hunk 3]
```diff
@@ -34,6 +36,10 @@ func (s StatusCode) IsError() bool {
 	return s == Error
 }
 
+func (s StatusCode) IsUpdateFailure() bool {
+	return s == Error
```
[the remark this answers is not in the document]
Comment (kubadz) [reply components/api-controller/pkg/apis/gateway.kyma.cx/meta/v1/types.go]: You are right with the first one. `IsEmpty` function seems correct.

File: components/api-controller/pkg/controller/networking/v1/networkingv1.go [hunk 4]
```diff
@@ -18,23 +20,41 @@ import (
...
 	return &istioImpl{
 		istioNetworkingInterface: i,
+		kubernetesInterface:      k,
 		istioGateway:             istioGateway,
 	}
 }
 
+type HostnameNotAvailableError struct {
+	errorMsg string
+}
```
[the remark this answers is not in the document]
Comment (kubadz) [reply components/api-controller/pkg/controller/networking/v1/networkingv1.go]: Good idea. Done.

File: components/api-controller/README.md [hunk 5]
```diff
@@ -37,12 +37,13 @@ $ go run cmd/controller/main.go
 
 Use these environment variables to configure the application:
 
-| Name | Required | Default | Description |
-|-----|---------|--------|------------|
-| API_CONTROLLER_LOG_LEVEL | No | `info` | Show detailed logs in the application.
-| DEFAULT_ISSUER | Yes | - | Used to set default issuer in NetworkPolicy
-| DEFAULT_JWKS_URI | Yes | - | Used to set default jwksUri in NetworkPolicy
-| GATEWAY_FQDN | Yes | - | Used to set gateway in VirtualServices spec
+| Name | Required | Default | Description | Possible Values |
```
[the remark this answers is not in the document]
Comment (kubadz) [reply components/api-controller/README.md]: Done.

File: components/api-controller/pkg/apis/gateway.kyma.cx/meta/v1/types.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sjanota) [reply components/api-controller/pkg/apis/gateway.kyma.cx/meta/v1/types.go]: My bad, you're right.

File: components/api-controller/README.md (hunk 1 above)
Comment (kubadz) [reply components/api-controller/README.md]: Done.

File: components/api-controller/README.md (hunk 2 above)
Comment (kubadz) [reply components/api-controller/README.md]: Done.

File: components/api-controller/pkg/controller/networking/v1/networkingv1.go [hunk 6]
```diff
@@ -157,3 +190,35 @@ func gatewayResourceFrom(vscv *istioNetworkingApi.VirtualService) *kymaMeta.Gate
...
+
+		for _, vs := range vsList.Items {
+			for _, occupiedHostname := range vs.Spec.Hosts {
+				if occupiedHostname == hostname {
+					//if hostname is not used by virtualservice actually assigned to CR
+					if vs.UID != assignedVirtualServiceUID {
+						return false, nil
+					}
+
+				}
```
Comment (piotrmsc) [inline components/api-controller/pkg/controller/networking/v1/networkingv1.go]: return here, we don;t need to iterate more

File: components/api-controller/pkg/controller/networking/v1/networkingv1.go [hunk 7]
```diff
@@ -157,3 +190,35 @@ func gatewayResourceFrom(vscv *istioNetworkingApi.VirtualService) *kymaMeta.Gate
...
+		vsList, err := a.istioVirtualServiceInterface(ns.GetName()).List(k8sMeta.ListOptions{})
+
+		if err != nil {
+			return false, err
+		}
+
+		for _, vs := range vsList.Items {
+			for _, occupiedHostname := range vs.Spec.Hosts {
+				if occupiedHostname == hostname {
+					//if hostname is not used by virtualservice actually assigned to CR
```
Comment (piotrmsc) [inline components/api-controller/pkg/controller/networking/v1/networkingv1.go]: `					//if hostname is used by virtualservice actually assigned to CR
					if vs.UID == assignedVirtualServiceUID {
						return true, nil
					}
                                          return false, nil
`

File: components/api-controller/pkg/controller/networking/v1/networkingv1.go [hunk 8]
```diff
@@ -157,3 +190,35 @@ func gatewayResourceFrom(vscv *istioNetworkingApi.VirtualService) *kymaMeta.Gate
 		Version: vscv.ResourceVersion,
 	}
 }
+
+func (a *istioImpl) isHostnameAvailable(metaDto meta.Dto, hostname string, assignedVirtualServiceUID types.UID) (bool, error) {
+
```
Comment (piotrmsc) [inline components/api-controller/pkg/controller/networking/v1/networkingv1.go]: ensure hostname is not empty - ie. case with kubectl edit without validate flag

File: components/api-controller/pkg/controller/networking/v1/networkingv1.go (hunk 6 above)
Comment (kubadz) [reply components/api-controller/pkg/controller/networking/v1/networkingv1.go]: Done.

File: components/api-controller/pkg/controller/networking/v1/networkingv1.go (hunk 7 above)
Comment (kubadz) [reply components/api-controller/pkg/controller/networking/v1/networkingv1.go]: Done, as it is the same suggestion as in the previous comment.

**Pull request merged**